editor/animation/animation_track_editor.cpp

```diff
@@ -1914,22 +1925,40 @@ void AnimationTimelineEdit::gui_input(const Ref<InputEvent> &p_event) {
 		accept_event();
 	}
 
-	if (mb.is_valid() && mb->is_pressed() && mb->get_button_index() == MouseButton::LEFT && hsize_rect.has_point(mb->get_position())) {
-		dragging_hsize = true;
-		dragging_hsize_from = mb->get_position().x;
-		dragging_hsize_at = name_limit;
+	if (!resizing_timeline && !dragging_hsize && mb.is_valid() && mb->is_pressed() && mb->get_button_index() == MouseButton::LEFT) {
+		if (timeline_resize_rect.has_point(mb->get_position())) {
+			resizing_timeline = true;
+			timeline_resize_from = timeline_resize_rect.position.x + timeline_resize_rect.size.width / 2 - mb->get_position().x;
+			timeline_resize_at = animation->get_length();
+		} else if (hsize_rect.has_point(mb->get_position())) {
+			dragging_hsize = true;
+			dragging_hsize_from = mb->get_position().x;
+			dragging_hsize_at = name_limit;
+		}
 	}
 
-	if (mb.is_valid() && !mb->is_pressed() && mb->get_button_index() == MouseButton::LEFT && dragging_hsize) {
-		dragging_hsize = false;
+	if (!resizing_timeline && mb.is_valid() && !panner->is_panning() && mb->get_button_index() == MouseButton::LEFT &&
+			mb->get_position().x > get_name_limit() && mb->get_position().x < (get_size().width - get_buttons_width())) {
+		int x = mb->get_position().x - get_name_limit();
+
+		float ofs = x / get_zoom_scale() + get_value();
+		emit_signal(SNAME("timeline_changed"), ofs, mb->is_alt_pressed());
+		dragging_timeline = true;
 	}
-	if (mb.is_valid() && mb->get_position().x > get_name_limit() && mb->get_position().x < (get_size().width - get_buttons_width())) {
-		if (!panner->is_panning() && mb->get_button_index() == MouseButton::LEFT) {
-			int x = mb->get_position().x - get_name_limit();
 
-			float ofs = x / get_zoom_scale() + get_value();
-			emit_signal(SNAME("timeline_changed"), ofs, mb->is_alt_pressed());
-			dragging_timeline = true;
+	if (mb.is_valid() && !mb->is_pressed() && mb->get_button_index() == MouseButton::LEFT) {
+		if (resizing_timeline) {
+			resizing_timeline = false;
+
+			if (animation->get_length() != timeline_resize_at) {
+				EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+				undo_redo->create_action(TTR("Change Animation Length"));
+				undo_redo->add_do_method(animation.ptr(), "set_length", animation->get_length());
+				undo_redo->add_undo_method(animation.ptr(), "set_length", timeline_resize_at);
+				undo_redo->commit_action();
```

Comment:
The animation seems to be continuously resized, so this can be

```suggestion
				undo_redo->commit_action(false);
```
I think.